hpx/util/pack_traversal.hpp

```diff
@@ -0,0 +1,729 @@
+
+        public:
+            explicit mapping_helper(M mapper)
+              : mapper_(std::move(mapper))
+            {
+            }
+
+            /// \copybrief try_traverse
+            template <typename T>
+            auto init_traverse(Strategy strategy, T&& element)
+                -> decltype(std::declval<mapping_helper>().try_traverse(
+                    strategy, std::declval<T>()))
+            {
+                return try_traverse(strategy, std::forward<T>(element));
+            }
+
+            /// Calls the traversal method for every element in the pack,
+            /// and returns a tuple containing the remapped content.
+            template <typename First, typename Second, typename... T>
+            auto init_traverse(strategy_remap_tag strategy, First&& first,
+                Second&& second, T&&... rest)
+                -> decltype(util::make_tuple(
+                    std::declval<mapping_helper>().try_traverse(
+                        strategy, std::forward<First>(first)),
+                    std::declval<mapping_helper>().try_traverse(
+                        strategy, std::forward<Second>(second)),
+                    std::declval<mapping_helper>().try_traverse(
+                        strategy, std::forward<T>(rest))...))
+            {
+                return util::make_tuple(
+                    try_traverse(strategy, std::forward<First>(first)),
+                    try_traverse(strategy, std::forward<Second>(second)),
+                    try_traverse(strategy, std::forward<T>(rest))...);
+            }
+
+            /// Calls the traversal method for every element in the pack,
+            /// without preserving the return values of the mapper.
+            template <typename First, typename Second, typename... T>
+            void init_traverse(strategy_traverse_tag strategy, First&& first,
+                Second&& second, T&&... rest)
+            {
+                try_traverse(strategy, std::forward<First>(first));
+                try_traverse(strategy, std::forward<Second>(second));
+                int dummy[] = {0,
+                    ((void) try_traverse(strategy, std::forward<T>(rest)),
+                        0)...};
+                (void) dummy;
+            }
+        };
+
+        /// Traverses the given pack with the given mapper and strategy
+        template <typename Strategy, typename Mapper, typename... T>
+        auto apply_pack_transform(
+            Strategy strategy, Mapper&& mapper, T&&... pack)
+            -> decltype(std::declval<mapping_helper<Strategy,
+                            typename std::decay<Mapper>::type>>()
+                            .init_traverse(strategy, std::forward<T>(pack)...))
+        {
+            mapping_helper<Strategy, typename std::decay<Mapper>::type> helper(
+                std::forward<Mapper>(mapper));
+            return helper.init_traverse(strategy, std::forward<T>(pack)...);
+        }
+    }    // end namespace detail
```

Comment:
would it make sense to move the stuff in detail into a seperate file? That way users could directly see the functions to use and not get hung up with all the implementation details.